### PR TITLE
BUG: Fix partition and argpartition error for empty input. Closes #6530

### LINKED ISSUE #6530: 1.10 regression: partition errors out on empty input
maintenance/1.9.x:
```
In [8]: ap = np.array([0, 2, 4, 6, 8, 10])

# Passing an empty 'k' array to partition is a no-op:
In [9]: ap.partition(np.array([], dtype=np.int64))

In [10]: ap
Out[10]: array([ 0,  2,  4,  6,  8, 10])
```

1.10.1:
```
In [2]: ap = np.array([0, 2, 4, 6, 8, 10])

In [3]: ap.partition(np.array([], dtype=np.int64))
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
<ipython-input-3-c708b3f943a6> in <module>()
----> 1 ap.partition(np.array([], dtype=np.int64))

MemoryError: 
```

Noticed this because it's breaking some tests in patsy.

### PR DESCRIPTION
**Pull request by yashmehrotra**
Fixes #6530 .

### CODE REVIEW

**Comment by yashmehrotra** [comment]
@njsmith @jaimefrio It is throwing a deprecation warning in one test case. What should I do ? I had 2 things in mind, either use `return 0` where the error is raised (in [here] (https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/item_selection.c#L1163) ) or add a check in test cases to skip the test for that particular environment.
What are your opinions ? Is there any better way ?

**Comment by yashmehrotra** [comment]
@njsmith  @jaimefrio Nevermind , its resolved. The problem was that the 32 bit system was trying to cast array into 64 bit. I used `dtype=int16` to avoid that.

**Comment by charris** [comment]
@jaimefrio This look good to you? The fix is as you suggested.

**Comment by jaimefrio** [comment]
There are `goto fail`s, at least in lines 832, 972, 981 and 988 that would need a `ret = -1` before them. Not sure if I am missing any other failure path that was relying on the initialization of `ret`...

**Comment by charris** [comment]
@jaimefrio Can you find time to put together a PR? @yashmehrotra No offence intended, but I want this fix soon for 1.10.2 and would prefer if someone familiar with the code got it done rather than go back and forth with corrections.

**Comment by yashmehrotra** [comment]
@charris It's ok, I understand, but I'd still commit the new changes that @jaimefrio mentioned. If it is working, we can merge it else I'll close this PR :smile: .

File: numpy/core/tests/test_item_selection.py [hunk 1]
```diff
@@ -68,6 +68,21 @@ def test_unicode_mode(self):
         k = b'\xc3\xa4'.decode("UTF8")
         assert_raises(ValueError, d.take, 5, mode=k)
 
+    def test_partition(self):
```
Comment (njsmith) [inline numpy/core/tests/test_item_selection.py]: should be called `test_empty_partition` to make clear what's special about this test

File: numpy/core/tests/test_item_selection.py [hunk 2]
```diff
@@ -68,6 +68,21 @@ def test_unicode_mode(self):
...
+        # In reference to github issue #6530
+        a = np.array([0, 2, 4, 6, 8, 10])
+        a = a.partition(np.array([], dtype=np.int16))
+
+        assert_equal(a, None)
```
Comment (njsmith) [inline numpy/core/tests/test_item_selection.py]: The above two lines don't make much sense -- `.partition` acts in-place and always returns `None`. It works to ensure that it isn't raising an error, but a better way to write the test would be:
```
a_original = np.array([0, 2, 4, 6, 8, 10])
a = a_original.copy()
# An empty partition should be a successful no-op
a.partition(np.array([], dtype=np.int16))
assert_equal(a, a_original)
```

Comment (njsmith) [inline numpy/core/tests/test_item_selection.py]: sorry, that should be `assert_array_equal` (and then you can remove the newly-added import of `assert_equal` at the top of the file)

**Comment by njsmith** [comment]
@charris: I did some careful review of the actual change to `item_selection.c`, and I'm fairly confident that it's correct. (The actual problem is simpler than I realized.) The tests need some tweaks as per above, but they're relatively trivial. Otherwise LGTM.

File: numpy/core/tests/test_item_selection.py [hunk 3]
```diff
@@ -68,6 +68,21 @@ def test_unicode_mode(self):
...
+
+    def test_argpartition(self):
```
Comment (njsmith) [inline numpy/core/tests/test_item_selection.py]: oh right, and this should be called `test_empty_argpartition`

File: numpy/core/tests/test_item_selection.py (hunk 2 above)
Comment (yashmehrotra) [inline numpy/core/tests/test_item_selection.py]: @njsmith I originally used the above method, don't know why I thought the current one would be better, anyway, I'll fix it now.
Thanks :)

**Comment by njsmith** [comment]
OK, one more comment about tests and then I promise I'm done :-). We should also the following to `numpy/core/tests/test_regression.py`:
```
    def test_empty_percentile(self):
        # gh-6530 / gh-6553
        assert_array_equal(np.percentile(np.arange(10), []), np.array([]))
```

because this was the call that I actually ran into the problem with (percentile calls partition internally). So it should be fixed by this, but we should add a test to make sure it stays fixed too.

**Comment by yashmehrotra** [comment]
@njsmith Oh, cool. I added the regression test. I hope its good to go now.

**Comment by njsmith** [comment]
LGTM, but I'll let @charris decide as release manager how he wants to handle this (presumably after confirming that the tests actually pass ;-)).

**Comment by yashmehrotra** [comment]
@njsmith Sure, No problem.

**Comment by charris** [comment]
@yashmehrotra Thanks, merged.

**Comment by jaimefrio** [comment]
Thanks for cleaning up my mess, all three of you!

**Comment by charris** [comment]
If you buy us beer and plane tickets we'll help you move ;)

**Comment by yashmehrotra** [comment]
+ 1. I am with charles on this one. :stuck_out_tongue:

**Comment by njsmith** [comment]
No no, you call it a "dev meeting" and get some company to pay for it ;-)

On Tue, Oct 27, 2015 at 9:17 PM, Charles Harris <notifications@github.com>
wrote:

> If you buy us beer and plane tickets we'll help you move ;)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6553#issuecomment-151717431>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

**Pull request merged**